### PR TITLE
vedit

### PR DESCRIPTION
**Pull request by vdike**
Added PBH realization functionality to realization_extensions class

### CODE REVIEW

File: pyHalo/realization_extensions.py [hunk 1]
```diff
@@ -166,3 +169,175 @@ def add_core_collapsed_halos(self, indexes, **kwargs_halo):
...
+            
+            _nsmooth = f_halos*rho_DM*(V/M)
+            _nclumpy = (1-f_halos)*rho_DM*(V/M)
+            
+            Nclumpy = int(np.random.poisson(_nclumpy))
+            Nsmooth = int(np.random.poisson(_nsmooth))
+            
+    
+            if Nsmooth > 0:
+                coord_x_smooth, coord_y_smooth,  z_shifts_smooth, masses_smooth = self._smooth_pbh (max_rendering_range, Nsmooth, center_x, center_y, zi, M)
```
Comment (dangilman) [inline pyHalo/realization_extensions.py:297]: regarding the comments above, this line should return the coord_x/y_smooth arrays, and the z_shifts_smooth can be defined in the line immediately following. For now, you can just create a list of masses, but if the future we might want to sample from a mass function (which is why you don't want to put that functionality where the halo coordinates are generated)

File: pyHalo/realization_extensions.py [hunk 2]
```diff
@@ -166,3 +169,175 @@ def add_core_collapsed_halos(self, indexes, **kwargs_halo):
...
+                pbh_x_coordinates = np.append(pbh_x_coordinates, coord_x_smooth)
+                pbh_y_coordinates = np.append(pbh_y_coordinates, coord_y_smooth)
+                pbh_redshifts += z_shifts_smooth
+                pbh_masses = np.append(pbh_masses, masses_smooth)
+            
+            # we can only generate PBH around halos if there are halos, hence the second condition
+            if Nclumpy > 0 and len(lens_model_list_at_plane): 
+                coord_x_clumpy, coord_y_clumpy, z_shifts_clumpy, masses_clumpy = self._clumpy_pbh (lens_model_list_at_plane, center_x, center_y,
```
Comment (dangilman) [inline pyHalo/realization_extensions.py:305]: same comment as above regarding the line following if Nsmooth > 0

File: pyHalo/realization_extensions.py [hunk 3]
```diff
@@ -166,3 +169,175 @@ def add_core_collapsed_halos(self, indexes, **kwargs_halo):
...
+                
+        return pbh_x_coordinates, pbh_y_coordinates, pbh_redshifts, pbh_masses
+    
+    def pbh_around_image(self, compfrac, M, image_index, realization, cosmology, lens_cosmo, cosmo_geometry, 
+                     zlens, zsource, x_interp_list, y_interp_list):
+    
+    
+        lens_plane_redshifts = self._realization.unique_redshifts
+        
+        max_rendering_range_base = 0.25 # arcsec
```
Comment (dangilman) [inline pyHalo/realization_extensions.py:320]: this should be an argument passed into the function

File: pyHalo/realization_extensions.py [hunk 4]
```diff
@@ -166,3 +169,175 @@ def add_core_collapsed_halos(self, indexes, **kwargs_halo):
...
+        npix = 300
```
Comment (dangilman) [inline pyHalo/realization_extensions.py:321]: should be able to pass this into function also

File: pyHalo/realization_extensions.py [hunk 5]
```diff
@@ -166,3 +169,175 @@ def add_core_collapsed_halos(self, indexes, **kwargs_halo):
...
+        
+        # defining things for later
+        conangle = cosmo_geometry.cone_opening_angle #arcsec 
+        LMF = LensingMassFunction(cosmology, zlens, zsource, cone_opening_angle=conangle)
+        mlow = 1e6
```
Comment (dangilman) [inline pyHalo/realization_extensions.py:326]: there are keyword arguments stored in the realization instance (self._realization) that define the mass range, access them with profile_args = self._realization._prof_params and then mlow = 10**profile_args['log_mlow'], etc.

File: pyHalo/realization_extensions.py [hunk 6]
```diff
@@ -166,3 +169,175 @@ def add_core_collapsed_halos(self, indexes, **kwargs_halo):
...
+                                                                cosmology, cosmo_geometry, max_rendering_range_base, 
+                                                                x_interp_list, y_interp_list, image_index, LMF, mlow, mhigh, rho_DM, npix, M, self._realization)
+                
+        # r3d is the three dimensional position of a subhalo which is used to compute a truncation radius, 
+        # this is irrelevant for PBH so we can just specify it with a dummy array of ones
+        r3d = np.array([None] * len(pbh_masses))
+        mdefs = ['PT_MASS'] * len(pbh_masses)
+        sub_flags = [False]*len(pbh_masses)
+    #    print('realization ' +str(image_index)+' contains '+str(len(pbh_masses))+ ' black holes')
+        profile_args = self._realization._prof_params
```
Comment (dangilman) [inline pyHalo/realization_extensions.py:340]: see above

**Review by dangilman — CHANGES_REQUESTED** [review]
thanks for pushing this! 

Most of my comments have to do with reorganizing the structure. If a function doesn't strictly have anything to do with an instance of Realization, it should be in utilities.py. Also, the function should be as general as possible, i.e. if it just generates points in a circle "pbh" should be left out of the function name (this makes the code a bit more readable, in my opinion).

Two other things: 1) eventually you'll need to write test functions for each routine, so think about how you can break the task down further into smaller bits that are each easy to test.
2) Add documentation to each function following the style in pyHalo (see for example find_core_collapsed_halos). PyCharm generates the documentation structure for you automatically, maybe Spider can do the same

File: pyHalo/realization_extensions.py [hunk 7]
```diff
@@ -166,3 +169,175 @@ def add_core_collapsed_halos(self, indexes, **kwargs_halo):
...
+            # keep circular symmetry
+            r = np.sqrt(x_sample_arcsec ** 2 + y_sample_arcsec ** 2)
+            keep = np.where(r <= Rmax)
+            x_out = np.append(x_out, x_sample_arcsec[keep])
+            y_out = np.append(y_out, y_sample_arcsec[keep])
+            
+        return x_out + x_0, y_out + y_0
+    
+        
+    def _smooth_pbh(self, max_rendering_range, Nsmooth, center_x, center_y, zi, M):
```
[the remark this answers is not in the document]
Comment (vdike) [reply pyHalo/realization_extensions.py:226]: I don't understand why I don't need to pass M and zi into the function when those are variables used in the function. Will it just pick those out of the namespace where they are defined?

[the remark this answers is not in the document]
Comment (vdike) [reply pyHalo/realization_extensions.py:226]: Sorry, I see what you mean: that part can be removed from the fn entirely

**Comment by vdike** [comment]
I added another commit with the improvements requested.

**Pull request closed without merging**